programs/whirlpool/src/state/tick.rs

```diff
@@ -752,3 +752,86 @@ mod array_update_tests {
         assert_eq!(*result, expected);
     }
 }
+
+#[cfg(test)]
+mod data_layout_tests {
+    use super::*;
+
+    #[test]
+    fn test_tick_array_data_layout() {
+        let tick_array_start_tick_index = 0x70e0d0c0i32;
+        let tick_array_whirlpool = Pubkey::new_unique();
+
+        let tick_initialized = true;
+        let tick_liquidity_net = 0x11002233445566778899aabbccddeeffi128;
+        let tick_liquidity_gross = 0xff00eeddccbbaa998877665544332211u128;
+        let tick_fee_growth_outside_a = 0x11220033445566778899aabbccddeeffu128;
+        let tick_fee_growth_outside_b = 0xffee00ddccbbaa998877665544332211u128;
+        let tick_reward_growths_outside = [
+            0x11223300445566778899aabbccddeeffu128,
+            0x11223344005566778899aabbccddeeffu128,
+            0x11223344550066778899aabbccddeeffu128,
+        ];
+
+        // manually build the expected Tick data layout
+        let mut tick_data = [0u8; Tick::LEN];
+        let mut offset = 0;
+        tick_data[offset] = tick_initialized as u8;
+        offset += 1;
+        tick_data[offset..offset + 16].copy_from_slice(&tick_liquidity_net.to_le_bytes());
+        offset += 16;
+        tick_data[offset..offset + 16].copy_from_slice(&tick_liquidity_gross.to_le_bytes());
+        offset += 16;
+        tick_data[offset..offset + 16].copy_from_slice(&tick_fee_growth_outside_a.to_le_bytes());
+        offset += 16;
+        tick_data[offset..offset + 16].copy_from_slice(&tick_fee_growth_outside_b.to_le_bytes());
+        offset += 16;
+        for i in 0..NUM_REWARDS {
+            tick_data[offset..offset + 16]
+                .copy_from_slice(&tick_reward_growths_outside[i].to_le_bytes());
+            offset += 16;
+        }
+
+        // manually build the expected TickArray data layout
+        // note: no discriminator
```

Comment:
any reason to not test with the discriminator/with `TickArray::try_deserialize` similar to the other tests?

Reply:
Good point!
Yep, there is a clear reason.

TickArray account doesn't implement `try_deserialize` because it is implemented by `AccountDeserialize` trait.
`AccountDeserialize` trait will be added by using `account` attribute macro (e.g. Whirlpool account).
https://docs.rs/anchor-lang/latest/anchor_lang/attr.account.html

But TickArray account uses `zero_copy` and it doesn't add `AccountDeserialize` trait.
So we need to test data layout in AccountLoader way. (bytemuck).
https://docs.rs/anchor-lang/latest/anchor_lang/attr.zero_copy.html

`bytemuck::from_bytes` (in `load()`) will be applied to [8..], so to test the validity to data layout mapping, first 8 bytes is not needed.
https://docs.rs/anchor-lang/latest/src/anchor_lang/accounts/account_loader.rs.html#165
```
        Ok(Ref::map(data, |data| {
            bytemuck::from_bytes(&data[8..mem::size_of::<T>() + 8])
        }))
```

Reply:
Ah, sorry if I'm misunderstanding, but I think `[account(zero_copy)]` will still generate `AccountDeserialize` (but not `AccountSerialize`, `AnchorSerialize` or `AnchorDeserialize`)

The `AccountDeserialize` it generates does the above + the discriminator checking ?
https://github.com/coral-xyz/anchor/blob/v0.29.0/lang/attribute/account/src/lib.rs#L180

Totally okay to keep as is though, since it's eventually how we load the TickArrays anyways in `peek_tick_array` https://github.com/orca-so/whirlpools/blob/main/programs/whirlpool/src/util/sparse_swap.rs#L290

Reply:
oh sorry, I overlooked that it still implement AnchorDeserialize 🤯
Yeah, I agree that we should check bytemuck instead of borsh. 👍